### PR TITLE
feat(BlockchainTree): ShareableBlockchainTree and pending state

### PR DESCRIPTION
**Pull request by rakita**
Create shareable blockchain tree that holds Tree behind RwLock.

Introduce traits
* `BlockchainTreeEngine` for write operations used by engine
* `BlockchainTreeViewer` for read operations and getting info data
* `BlockchainTreePendingStateProvider` for creating pending state data needed to execute on

Removes dependency from the engine on the provider and executor.

Refactor the Tree so the pending state data can be extracted and used with the state provider factroy.

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
I think I can follow and this makes sense to me.

So we have the `ShareableBlockchainTree` which clones the `PostSate` which is then used to create a `StateProvider` (the pending state), via the `canonical fork` block.

I believe we should rethink the `ShareableDatabase` type and merge with `BlockchainTreePendingStateProvider` so that we don't need two different types to construct the pending state and can directly call:

`StateProviderFactory::pending(&self)`

this would make it so much easier in rpc

File: crates/storage/provider/src/traits/state.rs [hunk 1]
```diff
@@ -71,21 +71,45 @@ pub trait StateProvider: BlockHashProvider + AccountProvider + Send + Sync {
...
 pub trait StateProviderFactory: Send + Sync {
-    /// History State provider.
-    type HistorySP<'a>: StateProvider
-    where
-        Self: 'a;
-    /// Latest state provider.
-    type LatestSP<'a>: StateProvider
-    where
-        Self: 'a;
-
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/storage/provider/src/traits/state.rs:82]: Yeah, I overcomplicated it initially with associated types, the idea was to have clear types but usage got complex.

File: crates/storage/provider/src/traits/state.rs [hunk 2]
```diff
@@ -71,21 +71,45 @@ pub trait StateProvider: BlockHashProvider + AccountProvider + Send + Sync {
...
     /// Storage provider for latest block.
-    fn latest(&self) -> Result<Self::LatestSP<'_>>;
+    fn latest<'a>(&'a self) -> Result<Box<dyn StateProvider + 'a>>;
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/storage/provider/src/traits/state.rs:75]: same thing, just less verbose.
using box dyn here is perfectly fine though

File: crates/storage/provider/src/traits/state.rs [hunk 3]
```diff
@@ -71,21 +71,45 @@ pub trait StateProvider: BlockHashProvider + AccountProvider + Send + Sync {
...
+    /// Return state provider over pending state.
+    fn pending_state_provider(
+        &self,
+        block_hash: BlockHash,
+    ) -> Result<Box<dyn PostStateDataProvider>>;
+}
+
+/// Post state data needs for execution on it.
+#[auto_impl[Box,&]]
+pub trait PostStateDataProvider: Send + Sync {
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/storage/provider/src/traits/state.rs]: copilot is amazing help for writing docs.

**Review by mattsse — APPROVED** [review]
🔥
lgtm

**Review by mattsse — APPROVED** [review]

**Pull request merged**